### ISSUE TITLE
Can't log in to Camdram from a linked OAuth account that uses a different email address

### ISSUE THREAD

**Issue by philosophicles** [camdram/camdram#462]
I have a Camdram account using `me@domain1.com` as my "official" Camdram email address (stored in the `users` table). 

This account is linked to all three of my Facebook, Google and Twitter accounts. My account settings confirms this to be the case. 

My Facebook and Twitter accounts also have `me@domain1.net` as my email address. My Google account is for a different email address, `myself@domain2.me.uk`. 

If I start on www.camdram.net homepage having explicitly logged out, and choose to log in with either Facebook or Twitter, the OAuth login flow works just fine and I land back on the homepage, visibly logged in correctly. 

However, repeating the above but logging in with Google, I am routed to Camdram's 'Register with Camdram' page. Here I am prompted to either create a new account (the two fields, name and email, are both pre-populated from Google), or to log in using an existing email and password combination (email is pre-propulated from Google) so I "can manually connect your Google and Camdram accounts". 

This is a bug - my Google and Camdram accounts are already manually connected so I shouldn't see this screen at all. 

(Manually logging in with existing username and password on the above page _does_ work, by the way.)

**Comment by CHTJonas** [comment]
How recently have you linked your accounts? Reason I ask is that I linked my Google account a while back and the `username` field in `acts_external_users` is completely numeric, whereas yours is an actual email address (hope you don't mind me searching you). I wonder if Google have changed their OAuth API or something.

Not sure if this is the reason but I also can't reproduce the behaviour you're observing...

**Comment by philosophicles** [comment]
Ah, that's a good thought. Actually, when I searched in `acts_external_users` for my surname, I got four records, two of which were for Google, but the one with an actual email in `username` was linked to my `user_id` while the other (with long numeric `username`) had `user_id = null`. 

So I guess there's kinda-sorta a bug here in that a row was entered into `acts_external_users` with null `user_id` - I think that's probably an orphan record that either shouldn't have been allowed in, or should have been cleaned up already. 

To answer your question directly - I've not had Google linked all that long, IIRC, I only had Facebook linked for a long time. But I'm not totally sure. 

I tried unlinking my Google account via the front-end and it removed the row with an actual email address, but not the one with null `user_id` (unsurprisingly). 

Re-linking to my Google account then succeeded, and updated the null `user_id` rather than inserting a new row (good). 

So I now have just one row for Google, and it is linked to my user ID, and the username is entirely numeric. And I can now log in successfully with Google. 

**TL;DR solved by unlinking and relinking external account via normal front-end account settings.**

**Comment by philosophicles** [comment]
I think you're probably right @CHTJonas that an API change has probably caused this. The weird row with `user_id = null` is bothersome but if I'm an isolated example, not worth any effort to track down root cause. I've documented the above so someone can find this again in the future if it happens again.  Since any user can fix this themselves through the front-end, I think that's sufficient and I'm happy to close this.

**Comment by CHTJonas** [comment]
Have tagged as 'wontfix', partially so it won't show up in the next set of release notes but also since the  original underlying bug that caused the null entry hasn't really been fixed (yet).

**Comment by CHTJonas** [comment]
The above being said, I've just run `SELECT * FROM acts_external_users WHERE user_id IS NULL ORDER BY last_login_at DESC` on the database and it's quite revealing. There are 265 of such null records with the last login being around 20th June. I wonder if something changed in release 20180616? @hoyes can you maybe shed any light on this? I guess we could DROP all those records from the DB but would be preferable to address the bug if possible.

**Comment by philosophicles** [comment]
It's occurred to me that this might not be a bug - perhaps this is how it's meant to work when users create new Camdram accounts using OAuth rather than with a Camdram-specific password? I.e. no row ever gets created in `acts_users` in such cases hence no link needed? @hoyes  can probably confirm. 

(I presume a row in `acts_users` would have to be created later if the user decides to link a second OAuth provider to their account. To relate the multiple rows in `acts_external_users` to each other.

If that's true, we might need to make a few more minor DB changes, since there are two columns on `acts_users` that aren't on `acts_external_users` which we might want to have for users who only have one OAuth provider and no Camdram password: 

1. `registered_at` (row creation timestamp - generally useful, I think we'd regret not having this sooner or later)
2. `is_email_verified` (relates to various other open issues I think)

Both of these are useful for each individual `external` row in isolation, regardless of whether a group of `external` rows relate to the same actual human, and both have relevance to `external` rows separate from `users` rows. 

Off the top of my head I'd reckon `acts_users.registered_at` would want to be set as the earliest related value from `acts_external_users`, if a `users` row is created later than the `external` row (rather than the timestamp at the moment the `users` row is created). 

I realize this issue isn't the right place for the above if it's a valid new task, but I thought it makes sense to get agreement here before raising a new issue... 

Tagging @GKFX as well as this might end up relating to the DB-improvement master plan...
